CMakeLists.txt

```diff
@@ -152,21 +157,7 @@ install(EXPORT      ${PROJECT_NAME}
         NAMESPACE   ${PROJECT_NAME}::
         DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME})
 
-set(PACKAGE_NAME ${PROJECT_NAME})
-set(prefix ${CMAKE_INSTALL_PREFIX})
-set(exec_prefix ${CMAKE_INSTALL_PREFIX})
-if (IS_ABSOLUTE "${CMAKE_INSTALL_INCLUDEDIR}")
-  set(includedir "${CMAKE_INSTALL_INCLUDEDIR}")
-else()
-  set(includedir "\${prefix}/${CMAKE_INSTALL_INCLUDEDIR}")
-endif()
-if (IS_ABSOLUTE "${CMAKE_INSTALL_LIBDIR}")
-  set(libdir "${CMAKE_INSTALL_LIBDIR}")
-else()
-  set(libdir "\${exec_prefix}/${CMAKE_INSTALL_LIBDIR}")
-endif()
```

Comment:
It's not clear to me how it went wrong in NixOS, unless `IS_ABSOLUTE` in cmake itself is broken? Is NixOS setting a bad value for -DCMAKE_INSTALL_LIBDIR and -DCMAKE_INSTALL_INCLUDEDIR?

If these are set to "lib" and "include" respectively, then NixOS should generate pkg-config files with the same logic both before and after this PR.

If these are set to absolute paths, then NixOS should generate absolute paths already.

I'm not even sure how it's *possible* to end up with:
```
prefix=/nix/store/3b25lbw0vnglfar91mich3mliqchxmiy-cglm-0.8.5
exec_prefix=/nix/store/3b25lbw0vnglfar91mich3mliqchxmiy-cglm-0.8.5
libdir=${prefix}//nix/store/3b25lbw0vnglfar91mich3mliqchxmiy-cglm-0.8.5/lib
includedir=${prefix}//nix/store/3b25lbw0vnglfar91mich3mliqchxmiy-cglm-0.8.5/include
```

This would imply that prefix= sees `-DCMAKE_INSTALL_PREFIX=/nix/store/hashed-dir-cglm-0.8.5` but that libdir and includedir see CMAKE_INSTALL_PREFIX with an ending slash, and see `-DCMAKE_INSTALL_LIBDIR=nix/store/hashed-dir-cglm-0.8.5/include` ***without*** a leading slash? Alternatively, `if (IS_ABSOLUTE "/nix/store/hashed-dir-cglm-0.8.5")` reports that that path is NOT an absolute path, which I simply do not understand.

@jtojnar is both a Nix person and the author of https://github.com/jtojnar/cmake-snips, which is a cmake function library that implements effectively the same logic as this PR moves away from. The README for the snip basically says "people frequently break Nix by not using this snip" :) so I think that NixOS probably specifically wants to avoid this PR after all?

But maybe @jtojnar can help shed light on why the NixOS package for cglm was getting weird pkg-config files.

Reply:
these are the cmake flags for cglm

```
cglm> cmake flags: 
-DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF 
-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON 
-DCMAKE_BUILD_TYPE=Release 
-DBUILD_TESTING=OFF 
-DCMAKE_INSTALL_LOCALEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/locale 
-DCMAKE_INSTALL_LIBEXECDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/libexec 
-DCMAKE_INSTALL_LIBDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib 
-DCMAKE_INSTALL_DOCDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/doc/cglm 
-DCMAKE_INSTALL_INFODIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/info 
-DCMAKE_INSTALL_MANDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/man 
-DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include 
-DCMAKE_INSTALL_INCLUDEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include 
-DCMAKE_INSTALL_SBINDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/sbin 
-DCMAKE_INSTALL_BINDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/bin 
-DCMAKE_INSTALL_NAME_DIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib 
-DCMAKE_POLICY_DEFAULT_CMP0025=NEW 
-DCMAKE_OSX_SYSROOT= 
-DCMAKE_FIND_FRAMEWORK=LAST 
-DCMAKE_STRIP=/nix/store/wn31i3dzwahz6ccws8bs1nwyqrpgsvj7-gcc-wrapper-11.3.0/bin/strip 
-DCMAKE_RANLIB=/nix/store/wn31i3dzwahz6ccws8bs1nwyqrpgsvj7-gcc-wrapper-11.3.0/bin/ranlib 
-DCMAKE_AR=/nix/store/wn31i3dzwahz6ccws8bs1nwyqrpgsvj7-gcc-wrapper-11.3.0/bin/ar 
-DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++
-DCMAKE_INSTALL_PREFIX=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5
```

Reply:
> Alternatively, `if (IS_ABSOLUTE "/nix/store/hashed-dir-cglm-0.8.5")` reports that that path is NOT an absolute path, which I simply do not understand.

So this is expanding out to:

```
if (IS_ABSOLUTE "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include")
  set(includedir "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include")
else()
  set(includedir "\${prefix}//nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include ")
endif()
if (IS_ABSOLUTE "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib")
  set(libdir "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib")
else()
  set(libdir "\${exec_prefix}//nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib")
endif()
```

Why is cmake claiming IS_ABSOLUTE is not, in fact, absolute? This doesn't make sense -- it directly contradicts CMake's documentation.

Reply:
Here's the cmake source for FileIsFullPath https://github.com/Kitware/CMake/blob/02599da236fd22db0dcfb6503194e5bad086aea9/Source/kwsys/SystemTools.cxx#L4262-L4293

Reply:
The quotes really should not matter, that's just cmake's optional syntax for strings.

Directories shouldn't matter either. I'm not using Nix.

Test file:
```cmake
project(test LANGUAGES C)

if (IS_ABSOLUTE "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include")
    message("literal is absolute")
else()
    message("literal is not absolute")
endif()

if (IS_ABSOLUTE "${CMAKE_INSTALL_INCLUDEDIR}")
    message("includedir variable as '${CMAKE_INSTALL_INCLUDEDIR}' is absolute")
else()
    message("includedir variable as '${CMAKE_INSTALL_INCLUDEDIR}' is not absolute")
endif()
```

```
$ cmake . -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include 
literal is absolute
includedir variable as '/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include' is absolute
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/cmtest
```

Reply:
And similarly, on my non-Nix system, I skipped the values for strip/ranlib/ar, but otherwise used your settings directly. I checked out commit a3186a7f32f174f8b4ff9a2961fd61f7dbd988d1 from before this merge.

```
+ cmake -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/doc/cglm -DCMAKE_INSTALL_INFODIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include -DCMAKE_INSTALL_SBINDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5 ..
```

```
-- The C compiler identification is GNU 11.1.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_TESTING
    CMAKE_CXX_COMPILER
    CMAKE_EXPORT_NO_PACKAGE_REGISTRY
    CMAKE_FIND_USE_PACKAGE_REGISTRY
    CMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY
    CMAKE_POLICY_DEFAULT_CMP0025


-- Build files have been written to: /tmp/cglm/cmbuild
```

Contents of cglm.pc
```
prefix=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5
exec_prefix=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5
libdir=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/lib
includedir=/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include

Name: cglm
Description: OpenGL Mathematics (glm) for C
URL: https://github.com/recp/cglm
Version: 0.8.6
Cflags: -I${includedir}
Libs: -L${libdir} -lcglm 
```

Reply:
I cannot reproduce the reported issue that inspired this PR. At all.

Reply:
Thank you for your investigation!

your test works for me too in nix develop
and the pc in /build contains the correct paths too in nix develop after running `cmakeConfigurePhase`
```
prefix=/home/artturin/cglm/outputs/out
exec_prefix=/home/artturin/cglm/outputs/out
libdir=/home/artturin/cglm/outputs/out/lib
includedir=/home/artturin/cglm/outputs/out/include
```

i made this nix reproducer which weirdly works

```nix
let
  pkgs = ((builtins.getFlake (toString ./.)).legacyPackages.${builtins.currentSystem});
  cmakefile = pkgs.writeText "CMakeLists.txt" ''
    project(test LANGUAGES C)

    if (IS_ABSOLUTE "/nix/store/rgfhd3jk5yflwcy112jwqvnx0dfw6bsr-cglm-0.8.5/include")
        message("literal is absolute")
    else()
        message("literal is not absolute")
    endif()

    if (IS_ABSOLUTE "''${CMAKE_INSTALL_INCLUDEDIR}")
        message("includedir variable as ' ''${CMAKE_INSTALL_INCLUDEDIR}' is absolute")
    else()
        message("includedir variable as ' ''${CMAKE_INSTALL_INCLUDEDIR}' is not absolute")
    endif()

    set(prefix ''${CMAKE_INSTALL_PREFIX})
    set(exec_prefix ''${CMAKE_INSTALL_PREFIX})
    if (IS_ABSOLUTE "''${CMAKE_INSTALL_INCLUDEDIR}")
      set(includedir "''${CMAKE_INSTALL_INCLUDEDIR}")
    else()
      set(includedir "\''${prefix}/''${CMAKE_INSTALL_INCLUDEDIR}")
    endif()
    if (IS_ABSOLUTE "''${CMAKE_INSTALL_LIBDIR}")
      set(libdir "''${CMAKE_INSTALL_LIBDIR}")
    else()
      set(libdir "\''${exec_prefix}/''${CMAKE_INSTALL_LIBDIR}")
    endif()

    configure_file(''${CMAKE_CURRENT_LIST_DIR}/test.pc.in ''${CMAKE_CURRENT_LIST_DIR}/test.pc @ONLY)
  '';
  pcfile = pkgs.writeText "test.pc.in" ''
    prefix=@prefix@
    exec_prefix=@exec_prefix@
    libdir=@libdir@
    includedir=@includedir@
  '';
in
pkgs.stdenv.mkDerivation {
  name = "cmake-test";
  nativeBuildInputs = [ pkgs.cmake ];
  buildCommand = ''
    cp ${cmakefile} CMakeLists.txt
    cp ${pcfile} test.pc.in
    cmakeConfigurePhase
    mkdir $out
    mv /build/* $out
  '';
}
```

Reply:
0.8.5 did not have the `is_absolute` check:

https://github.com/recp/cglm/blob/v0.8.5/CMakeLists.txt#L153-L154

Reply:
So no one tested cglm to see if it was already fixed. I see.

In fact, the right hand of NixOS reported the same issue in #249 and fixed it in #250 and then the left hand of NixOS reverted that fix.

Reply:
indeed https://github.com/recp/cglm/pull/272

Reply:
@eli-schwartz Thanks for bringing this up. This has been a while back and I can't really remember why I did this PR, but I clearly wasn't aware of the earlier fixes. I think I went overboard simplifying the cmake configuration. I need to learn more about cmake best practices. It has confused me in another project as well. If you have any sources on how to write proper cmake configuration I'm very interested.

Sorry for the confusion/problems this has caused. The revert seems like the best solution for now.